### PR TITLE
[DO NOT MERGE]Pm d0ix s2idle

### PR DESCRIPTION
**Pull request by keyonjie**
This is for s2idle KWD support.

    When system is entering into S0ix, we will put ADSP to D0i3 instead of
    D3, this will enable keeping wake on voice stream active and FW will
    send IPC interrupt to wake up the whole system when needed.

### CODE REVIEW

**Comment by keyonjie** [comment]
@plbossart @ranj063 it might be better to add some configure param to decide whether enter D0i3 or D3 when system is entering s2idle. But let me know if this sounds fine for you firstly.

File: sound/soc/sof/pm.c [hunk 1]
```diff
@@ -429,12 +429,61 @@ EXPORT_SYMBOL(snd_sof_runtime_resume);
...
+				ret);
+			return ret;
+		}
+
+		pci_restore_state(pci);
+		disable_irq_wake(pci->irq);
+
+		/* resume DMA trace, only need send ipc */
+		ret = snd_sof_init_trace_ipc(sdev);
+		if (ret < 0) {
```
Comment (ranj063) [inline sound/soc/sof/pm.c:481]: @keyonjie the sof_resume sequenc already includes resuming the trace DMA. Why do it again?

File: sound/soc/sof/pm.c [hunk 2]
```diff
@@ -429,12 +429,61 @@ EXPORT_SYMBOL(snd_sof_runtime_resume);
...
 }
 EXPORT_SYMBOL(snd_sof_resume);
 
 int snd_sof_suspend(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	int ret;
+
+	if (acpi_target_system_state() == ACPI_STATE_S0) {
```
Comment (ranj063) [inline sound/soc/sof/pm.c:502]: this is confusing. You check for the same state in suspend and resume?

File: sound/soc/sof/pm.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/pm.c:481]: oops sorry this is in sof_resume. Why do these things have to be conditional at all? resume would always be to enter S0 right?

File: sound/soc/sof/pm.c (hunk 2 above)
Comment (keyonjie) [reply sound/soc/sof/pm.c:502]: correct, for resuming, I should not use this, let me change.

File: sound/soc/sof/pm.c (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/pm.c:481]: yes, why we need to make it conditional is that for resuming from D0i3, we don't need to do the full resuming, as we haven't done those suspending at .suspend(). I will update my PR to reflect that.

**Comment by keyonjie** [comment]
@ranj063 updated.

**Comment by keyonjie** [comment]
I think this is aligned but let's merge it with the rest S0ix part in a short, so I merge this to PR#1249 and marking this with the prefix [DO NOT MERGE].

**Comment by plbossart** [comment]
> I think this is aligned but let's merge it with the rest S0ix part in a short, so I merge this to PR#1249 and marking this with the prefix [DO NOT MERGE].

If this PR is no longer useful or part of another, please close it. I don't understand what you are doing here.

**Comment by keyonjie** [comment]
@plbossart yes, closing now.

**Pull request closed without merging**